features/64-chat-offline-cache/api.md

```diff
@@ -0,0 +1,373 @@
+
+skygearChat.subscribeMessages(conversation: conversation,
+                              handler: func (messageDeltas) {
+                                this.messages = skygearChat.applyDelta(messageDelta, toList: this.messages)
+                                this.tableView.reloadTable()
+                              })
+```
+
+#### Fetch latest messages and update table
+
+When
+- enter conversation and try to skip to latest message
+
+```
+let conversation = this.conversation
+let beforeMessage = nil // fetch latest message
+let limit = 100
+let order = .descending // latest come first
+
+skygearChat.fetchMessages(conversation: conversation,
+                          beforeMessage: beforeMessage,
+                          limit: limit,
+                          order: order,
+                          completion: func (messages, cached = false) {
+                            this.messages = skygearChat.merge(messages, toList: this.messages)
+                            this.tableView.reloadTable()
+                          })
+```
+
+#### Fetch newer messages and update table
+
+When
+- enter conversation and try to stay at last message position and load newer message from that point
+- come offline and online again when in conversation
+- load newer messages when scroll to bottom
+
+```
+let conversation = this.conversation
+let afterMessage = this.latestMessage
+let limit = 100
+let order = .descending // latest come first
+
+skygearChat.fetchMessages(conversation: conversation,
+                          afterMessage: afterMessage,
+                          limit: limit,
+                          order: order,
+                          completion: func (messages, cached = false) {
+                            this.latestMessage = messages[0]
+                            this.messages = skygearChat.merge(messages, toList: this.messages)
+                            this.tableView.reloadTable()
+                          })
+```
+
+#### Fetch older messages and update table
+
+- load older messages when scroll to top
+
+```
+let conversation = this.conversation
+let beforeMessage = this.firstMessage
+let limit = 100
+let order = .descending // latest come first
+
+skygearChat.fetchMessages(conversation: conversation,
+                          beforeMessage: firstMessage,
+                          limit: limit,
+                          order: order,
+                          completion: func (messages, cached = false) {
+                            this.latestMessage = messages[0]
+                            this.messages = skygearChat.merge(messages, toList: this.messages)
+                            this.tableView.reloadTable()
+                          })
+```
+
+### With ManagedMessageList
+
+#### Enter conversation
+
+```
+this.messageList = new ManagedMessageList(conversation: this.conversation,
+                                          delegate: this)
+
+if (skygearChat.pubsub.isConnected) {
```

Comment:
I think ManagedMessageList should handle it's relation with pubsub internally, instead of asking Developers to write these few lines of code all the time?

Reply:
Then I think the `constructor` should also accept an instance of skygear container.

Reply:
Sure.